modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/rel/HashJoinNode.java

```diff
@@ -0,0 +1,623 @@
+        Collection<RowT> coll = Collections.emptyList();
+
+        for (Integer entry : leftJoinPositions) {
+            Object ent = handler.get(entry, row);
+
+            if (ent == null) {
+                return Collections.emptyList();
+            }
+
+            Object next0 = next.get(ent);
+
+            if (next0 == null) {
+                return Collections.emptyList();
+            }
+
+            processed++;
+            if (processed == leftJoinPositions.size()) {
+                coll = (Collection<RowT>) next.get(ent);
+
+                if (processTouched) {
+                    ((TouchedList<RowT>) coll).touched = true;
+                }
+            } else {
+                next = (Map<Object, Object>) next0;
+            }
+        }
+
+        return coll;
+    }
+
+    private static <RowT> List<RowT> getUntouched(Map<Object, Object> entries, @Nullable List<RowT> out) {
+        if (out == null) {
+            out = new ArrayList<>();
+        }
+
+        for (Map.Entry<Object, Object> ent : entries.entrySet()) {
+            if (ent.getValue() instanceof Collection) {
+                TouchedList<RowT> coll = (TouchedList<RowT>) ent.getValue();
+                if (!coll.touched) {
+                    out.addAll(coll);
+                }
+            } else {
+                getUntouched((Map<Object, Object>) ent.getValue(), out);
+            }
+        }
+        return out;
+    }
+
+    private static <RowT> void resetTouched(Map<Object, Object> entries) {
+        for (Map.Entry<Object, Object> ent : entries.entrySet()) {
+            if (ent.getValue() instanceof Collection) {
+                TouchedList<RowT> coll = (TouchedList<RowT>) ent.getValue();
+                if (coll.touched) {
+                    coll.touched = false;
+                }
+            } else {
+                resetTouched((Map<Object, Object>) ent.getValue());
+            }
+        }
+    }
+
+    @Override
+    protected void pushRight(RowT row) throws Exception {
+        assert downstream() != null;
+        assert waitingRight > 0;
+
+        checkState();
+
+        waitingRight--;
+
+        Map<Object, Object> next = hashStore;
+        int processed = 0;
+        for (Integer entry : rightJoinPositions) {
+            processed++;
+            Object ent = handler.get(entry, row);
+            if (processed == rightJoinPositions.size()) {
+                Collection<RowT> raw = touchResults
+                        ? (Collection<RowT>) next.computeIfAbsent(ent, k -> new TouchedList<>())
+                        : (Collection<RowT>) next.computeIfAbsent(ent, k -> new ArrayList<>());
+
+                raw.add(row);
+            } else {
+                next = (Object2ObjectOpenHashMap<Object, Object>) next.computeIfAbsent(ent, k -> new Object2ObjectOpenHashMap<>());
+            }
+        }
+
+        if (waitingRight == 0) {
+            rightSource().request(waitingRight = inBufSize);
+        }
+    }
```

Comment:
this looks overcomplicated. Is it possible to store everything in a single map?

Reply:
ok, i rewrite this implementation and found significant performance boost, lets stay with new implementation although it would be difficult to implement "is [not] distinct" and seems we need to implement different execution node for such a case.

new performance run:
```
Benchmark                      (clusterSize)  (fsync)  Mode  Cnt       Score      Error  Units
SqlJoinBenchmark.leftHashJoin              1    false  avgt   20  265234.571 ± 3363.419  us/op
```
previous is actual in PR header, possible i need to refresh it, after review was completed.